GLTFLoader v2

richtr opened the pull request:
This is a complete refactor/rewrite of the existing glTF code in https://github.com/mrdoob/three.js/tree/dev/examples/js/loaders/gltf.

The new `GLTFLoader.js` includes _all_ of the features of the legacy glTF code.

The benefits of this rewrite are as follows:
- added async asset loading (buffers, textures, images) to @mrdoob's rewritten [GLTFLoader.js](https://github.com/mrdoob/three.js/blob/e33a85ab1a2d0cf3661b2024f7828c380691a635/examples/js/loaders/GLTFLoader.js).
- better code readability and less code indirection
- more compact internal glTF object representations (use THREE.js native objects as early and often as possible)
- feature parity with the legacy glTF code (glTF v1.0 support w/ shaders, animations, skins, lights, extensions, etc) with a >60% reduction in size (1057 lines of code vs 2696 lines of code. 18KB vs 41KB when run through https://jscompress.com/)

richtr committed:
Initial GLTFLoader updates so it will also obtain async resources over the network

richtr committed:
First take at adding shader support to new GLTFLoader

richtr committed:
First working custom shaders version of GLTFLoader.js

richtr committed:
Fully working custom GLTF shaders + general library fix-ups

richtr committed:
Add skins, animations, lights, extensions and continue to refactor the library

richtr committed:
Refactor new GLTFLoader (seperate glTF parsing and internal library building)

richtr commented:
The objective here is to deprecate https://github.com/mrdoob/three.js/tree/dev/examples/js/loaders/gltf in favor of a single glTF loader in the THREE.js project.

We should eventually move https://github.com/mrdoob/three.js/tree/dev/examples/js/loaders/gltf to https://github.com/mrdoob/three.js/tree/e33a85ab1a2d0cf3661b2024f7828c380691a635/examples/js/loaders/deprecated, update the docs to point to the new class and update the glTF example to use the new class.

I feel this could be a separate pull request once this new GLTFLoader.js is merged.

/ping the original authors @tparisi @mrdoob for review.

richtr committed:
Fix GLSL shader update binding to point at the correct nodes + more general refactoring

richtr committed:
Make _each thenable and refactor dependencies graph + other minor fixes

richtr commented:
@mrdoob do you prefer everything squashed in to one commit or is the current commit history acceptable?

richtr commented:
Also, I'm wondering whether instead of adding `/examples/webgl_loader_gltf_new.html` whether this PR should instead fix `/examples/webgl_loader_gltf.html` to use this new loader (+ update the references in the GLTFLoader documentation). Or should that wait for a later commit?

tparisi commented:
@richtr this looks like a lovely refactor; but I haven't actually tried it - is there a fork I can grab to actually test everything out against it?

richtr commented:
> is there a fork I can grab to test everything out against it?

You can clone and switch to the branch that this is on:

``` bash
git clone git@github.com:richtr/three.js.git
git checkout feature/GLTFLoader
```

Then run a local web server using a [one-liner](https://gist.github.com/willurd/5720255) within the root directory.

I've also mirrored this branch to here: https://richtr.github.io/three.js/examples/webgl_loader_gltf_new.html if you want to see how it works quickly too.

I've tried the branch with all the models in https://github.com/KhronosGroup/glTF/tree/master/sampleModels but it would be good to get more testing in too.

fraguada commented:
@richtr I've been wanting migrate the json file I'm currently generating to glTF.  I've tested a few .dae files converted with https://cesiumjs.org/convertmodel.html with your feature/GLTFLoader branch and they seem to work fine.  This pull made me look into gltf and for a file that was 27,9mb with the Three.js format, it becomes 13,2mb in GLTF and 4,7mb zipped.  

Geometry loaded fine, but textures as jpgs did not load for my files.  I'm not sure if it is something in particular with my model.  I don't get any errors that the .pngs are not found, but the meshes don't get a texture applied.  Maybe a sample file generated from my process would help?

fraguada commented:
@richtr [This is a simple box](https://drive.google.com/file/d/0B6MVdY0n99OhR0p5UXlUTUE0SFk/view?usp=sharing).  It has a material with a diffuse channel as a jpg.  The box itself loads, but I get errors such as: 
`ERROR: Missing fragment shader definition: tbox0FS    GLTFLoader.js:1117:7
THREE.MeshPhongMaterial: 'uniforms' is not a property of this material.    three.js:7248:6
THREE.Material: 'fragmentShader' parameter is undefined.    three.js:7239:6
THREE.MeshPhongMaterial: 'vertexShader' is not a property of this material.    three.js:7248:6`

I've compared with the other glts files and while there are some differences, I am not sure which ones are causing the material to not be understood.  The program and shaders seems to be defined in the same manner, and the glsl files have the same relationship in the directory to the .gltf file.

The process to make this: Create box in Rhino 3d, export as .dae, run through collada2gltf with no special options, define the model in your `webgl_loader_gltf_new.html` and run the scene.  Other models show just fine.

Commit:
Improves spec compliance: primitive.mode is optional

AVGP commented:
Great work, @richtr!

I'm looking forward to see this land :+1: 

I also tested the loader with:

* Data URIs ✓
* Primitive without indices	✓
* Primitive without mode ✗ [Pull request pending]

I'll see if I can find some more test models :)

richtr committed:
Merge pull request #1 from archilogic-com/feature/GLTFLoader

Improves spec compliance: primitive.mode is optional

richtr commented:
Thanks for the pull request @AVGP! This has been merged in to this branch now.

@fraguada I'm yet to look in to this issue but I will endeavor to do so shortly!

mrdoob commented:
> @mrdoob do you prefer everything squashed in to one commit or is the current commit history acceptable?

No need to squash 😊

> Also, I'm wondering whether instead of adding `/examples/webgl_loader_gltf_new.html` whether this PR should instead fix `/examples/webgl_loader_gltf.html` to use this new loader (+ update the references in the GLTFLoader documentation). 

I would just fix  `/examples/webgl_loader_gltf.html`.

mrdoob commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -12,368 +16,1836 @@ THREE.GLTFLoader.prototype = {
 	},
 
-	parse: function ( json ) {
+	setBaseUrl: function( value ) {
```
I've been renaming this method to `setPath` in other loaders:

https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/OBJLoader.js#L56-L60

richtr replied:
It's good to have consistency here. Fixed in e0e7dbf9fefa8ee6a01e580ca0801fd2c9bf0f40.

tparisi commented:
I'm ok with all this... let er rip

mrdoob commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -12,368 +16,1836 @@ THREE.GLTFLoader.prototype = {
-			nodes: {},
-			scenes: {}
-		};
+THREE.GLTFShaders = new GLTFRegistry();
```
How about setting this in the loaders instead?

```js
THREE.GLTFLoader.Registry = new GLTFRegistry();
```

richtr replied:
Good idea and updated in e0e7dbf9fefa8ee6a01e580ca0801fd2c9bf0f40.

`THREE.GLTFShaders` -> `THREE.GLTFLoader.Shaders`
`THREE.GLTFAnimator` -> `THREE.GLTFLoader.Animations`

mrdoob commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -12,368 +16,1836 @@ THREE.GLTFLoader.prototype = {
+				if ( this.isRot ) {
 
-			var texture = textures[ textureId ];
+					this.quat1.set( this.values[ i * 4 ],
```
I think you should be able to do `this.quat1.fromArray( this.values, i * 4 );`

richtr replied:
Good tip. I went through the loader and fixed this wherever I could in e0e7dbf9fefa8ee6a01e580ca0801fd2c9bf0f40.

cx20 commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -12,368 +16,1836 @@ THREE.GLTFLoader.prototype = {
+										uvalue = value;
+										materialParams.transparent = true;
+
+									}
```
I think "else" statement is necessary here.

```javascript
if ( pname == "transparency" ) {
    uvalue = value;
    materialParams.transparent = true;
} else {
    uvalue = value;
}
```

See: https://github.com/mrdoob/three.js/issues/8381

richtr replied:
I've added a fix for #8381 in e0e7dbf9fefa8ee6a01e580ca0801fd2c9bf0f40.

Thanks @cx20.

cx20 replied:
Thank you for fix. However, it looks like the code https://github.com/mrdoob/three.js/commit/e0e7dbf9fefa8ee6a01e580ca0801fd2c9bf0f40 has not been fixed.
If you did not add the "else" statement, uvalue is undefined.

For example, following code was extracted from duck.gltf.

```javascript
    "materials": {
        "blinn3-fx": {
            "technique": "technique0",
            "values": {
                "shininess": 38.4,
            }
        }
    },
    "techniques": {
        "technique0": {
            "parameters": {
                "shininess": {
                    "type": 5126
                },
```
When you load the above file, it will be in the following result.

```javascript
case WEBGL_CONSTANTS.FLOAT: // 5126

    uvalue = shaderParam.value; // shaderParam.value is undefined (= technique.parameters["shininess"].value)

    if ( pname == "transparency" ) {
        uvalue = value;
        materialParams.transparent = true;
    } else { // please add the "else" statement
        uvalue = value;  // value is 38.4
    }

    break;
```

richtr replied:
You're absolutely right and this was an oversight on my part. 

I have (really truely!) added this now in 99bf046573c50240bc5adf99b7cb263e10511739 :)

cx20 replied:
Thanks!

richtr committed:
Update GLTFLoader v2 with feedback on PR: https://github.com/mrdoob/three.js/pull/9786

- Replace .set(...) with .fromArray(...) where possible for THREE.js Math objects (thanks @mrdoob)
- Include fix for https://github.com/mrdoob/three.js/issues/8381 (thanks @cx20)
- Move THREE.GLTFShaders to THREE.GLTFLoader.Shaders
- Move THREE.GLTFAnimator to THREE.GLTFLoader.Animations
- Rename .setBaseUrl(...) to .setPath(...)
- Update demo page 'examples/webgl_loader_gltf_new.html' with renamed APIs

richtr committed:
Replace 'examples/webgl_loader_gltf.html' with new version of GLTFLoader

richtr committed:
Update documentation at /docs/?q=gltf#Reference/Loaders/GLTFLoader to reference new GLTFLoader

richtr committed:
Move /examples/js/loaders/gltf/* to /examples/js/loaders/deprecated/gltf/*

richtr commented:
>> Also, I'm wondering whether instead of adding /examples/webgl_loader_gltf_new.html whether >> this PR should instead fix /examples/webgl_loader_gltf.html to use this new loader (+ update >> the references in the GLTFLoader documentation).
> 
> I would just fix /examples/webgl_loader_gltf.html.

OK. I have moved `/examples/webgl_loader_gltf_new.html` to `/examples/webgl_loader_gltf.html` (thereby overwriting the old loader demo). Fixed in ec544003ab85b3d7662451cc54bb75403b5fae11.

In addition, I have updated the documentation to reflect the API and point to this new GLTFLoader in commit 5b0bfba9f75574378e903522308f9263d7a007b4.

In db17203a2cd9fb155b4038e2b9127b1070a1c29d I have moved the old `/examples/js/loaders/gltf/*` to `/examples/js/loaders/deprecated/gltf/*` to avoid confusion on which glTF loader to use in the future. I'd be happy to revert this commit if we want to keep that around for a while.

AVGP commented:
@richtr It didn't seem to work with transparent KHR materials, but maybe I misread the spec when I did our exporter, but I *think* `transparent` is inside the `values` property and I could not find the loader setting `opacity` on the material - I attached a PR to your branch, please feel free to correct my assumptions if I went wrong somewhere.

richtr committed:
Actually add a fix for https://github.com/mrdoob/three.js/issues/8381

richtr committed:
Re-fix https://github.com/mrdoob/three.js/issues/8381 in a more logical way

richtr committed:
Update material parameter checking and validation.

richtr committed:
Fix issue whereby the spec says doubleSided and transparent should be in the 'values' section but every use of these params I've seen in the wild has been outside this section

richtr committed:
Ambient is not a valid properties for THREE.js materials

richtr committed:
THREE textures have flipY set to true by default. By setting to false we can remove double flip from GLTFLoader

richtr commented:
@fraguada wrote:
> @richtr [This is a simple box](https://drive.google.com/file/d/0B6MVdY0n99OhR0p5UXlUTUE0SFk/view?usp=sharing). It has a material with a diffuse channel as a jpg. The box itself loads, but I get errors such as: 
ERROR: Missing fragment shader definition: tbox0FS GLTFLoader.js:1117:7 THREE.MeshPhongMaterial: 'uniforms' is not a property of this material. three.js:7248:6 THREE.Material: 'fragmentShader' parameter is undefined. three.js:7239:6 THREE.MeshPhongMaterial: 'vertexShader' is not a property of this material. three.js:7248:6
>
> So I can see the box, but no material.

It seems fine to work with no errors for me with the 'View-Perspective' camera you have set up in your glTF file:

<img width="340" alt="screen shot 2016-10-04 at 1 50 05 am" src="https://cloud.githubusercontent.com/assets/497699/19058083/e84f7bec-89d4-11e6-9f70-de9263dbd221.png">

I see no errors on the console when I load this (both `glTF` and `glTF-MaterialsCommon` versions provided work). I couldn't really get the other (orthographic) cameras working. Are they set up correctly in the glTF file?

richtr committed:
Fix issue where Firefox Promise then callback can not correctly keep track of its key

richtr commented:
Actually @fraguada I should have asked which browser you were using. I _did_ get the errors on the console you also encountered above in the latest Firefox. This has been fixed with https://github.com/mrdoob/three.js/pull/9786/commits/8d254a4843e92588adc96f9c98c216a349e6230f.

richtr commented:
@mrdoob @tparisi I believe I've addressed all the feedback and comments on this PR to date. Ready to merge?

richtr commented:
@mrdoob When this lands I believe we can also close https://github.com/mrdoob/three.js/pull/9734 and https://github.com/mrdoob/three.js/pull/9803. Or they could be merged first, then this will deprecate those old interfaces.

cx20 commented:
@richtr GLTFLoader v2 does not support Safari?

I tested old and new glTF loaders.
https://github.com/cx20/gltf-test

@mrdoob's GLTFLoader (basic loader) can display a duck in Safari. 
However,  GLTFLoader v2 (new loader) could not display it in Safari.

tparisi commented:
Are we back to the Safari bug? There is an issue filed here which was recently closed

https://github.com/mrdoob/three.js/issues/8381

@richtr did you pick up those changes?

cx20 commented:
@tparisi I think that a closed Safari issue is #8973. 
#8381 is different from the Safari issue.

richtr commented:
> @richtr GLTFLoader v2 does not support Safari?

It does but there is an issue with glTF-Embedded demos in Safari (`glTF` and `glTF-MaterialsCommon` demos are fine). I'm hitting this issue: https://stackoverflow.com/questions/30920059/xhr-data-uri-doesnt-work-in-safari. 

We currently run everything (including data URIs) through THREE.XHRLoader to save on parallel code paths (one for data URIs and another for http or https URLs).

I'd prefer that this issue would be fixed in THREE.XHRLoader so it works everywhere else this could become an issue too.

cx20 commented:
@richtr v2 is supports Safari. However, `glTF-Embedded` is not supports in Safari. I understood.
I thought to be a problem of judgment rather than a technical problem.
I thought that I should port mrdoob's stringToArrayBuffer function simply. 
I think that is a good idea that an issue is fixed in `THREE.XHRLoader`.

richtr commented:
I've created a new pull request to fix [the issue documented above](https://github.com/mrdoob/three.js/pull/9786#issuecomment-251680940) in `THREE.XHRLoader` here: https://github.com/mrdoob/three.js/pull/9823.

Using https://github.com/mrdoob/three.js/pull/9823 and this PR (https://github.com/mrdoob/three.js/pull/9786) together fixes the issue of loading and rendering `glTF-Embedded` on Safari OS X and iOS.

richtr commented:
Ready to go @mrdoob @tparisi. Please also take a look at https://github.com/mrdoob/three.js/pull/9823.

richtr committed:
Remove debugging from examples/loaders/deprecated/gltf/glTFLoader.js

richtr committed:
Refactor waitForDependencies in to _withDependencies. Remove need for global library.* tracking. Other general refactoring

mrdoob commented:
Thaaanks! 😊

richtr commented:
Thanks @mrdoob! (many thanks to @tparisi, @cx20, @fraguada, @AVGP too!)
